hubblestack/hangtime/linux_itimers.py

```diff
@@ -90,21 +95,22 @@ def restore(self, ended=False):
         while self.prev and self in self.prev:
             self.prev.pop()
         if self.prev:
-            p = self.prev[-1]
-            dt = time.time() - p.started
-            tr = p.timeout - dt
-            if tr > 0:
-                log.debug("time remaining on previous timer %s tr=%f, restarting itimer", repr(p), tr)
-                signal.signal(signal.SIGALRM, p.fire_timer)
-                signal.setitimer(signal.ITIMER_REAL, tr)
+            prev = self.prev[-1]
+            time_diff = time.time() - prev.started
+            time_remaining = prev.timeout - time_diff
+            if time_remaining > 0:
+                log.debug("time remaining on previous timer %s tr=%f, restarting itimer",
+                          repr(prev), time_remaining)
+                signal.signal(signal.SIGALRM, prev.fire_timer)
+                signal.setitimer(signal.ITIMER_REAL, time_remaining)
             else:
-                p.restore()
+                prev.restore()
         else:
             log.debug("timer stack empty after %s, resetting signals/itimers", repr(self))
             signal.setitimer(signal.ITIMER_REAL, 0)
             signal.signal(signal.SIGALRM, signal.SIG_DFL)
 
-    def fire_timer(self, *sig_param):
```

Comment:
I don't think this will actually work. when signal() fires, it passes two arguments to the callback function. without either the two args or a star-args, it'll likely crash during the signal. Have you tested this?